Add a per-site toggle in the legacy identity and the new trust panel UI to add the site to the sanitizing exceptions
---
diff --git a/browser/base/content/browser-siteIdentity.js b/browser/base/content/browser-siteIdentity.js
index 5db8eb880c..a309d8647a 100644
--- a/browser/base/content/browser-siteIdentity.js
+++ b/browser/base/content/browser-siteIdentity.js
@@ -234,6 +234,9 @@ var gIdentityHandler = {
       "identity-popup-clear-sitedata-button": event => {
         this.clearSiteData(event);
       },
+      "identity-popup-allow-sitedata-button": () => {
+        this.toggleSiteData();
+      },
       "identity-popup-remove-cert-exception": () => {
         this.removeCertException();
       },
@@ -504,6 +507,33 @@ var gIdentityHandler = {
     event.stopPropagation();
   },
 
+  async refreshSiteData() {
+    document.getElementById("identity-popup-allow-sitedata-toggle").toggleAttribute(
+      "pressed",
+      Services.perms.testExactPermissionFromPrincipal(
+        gBrowser.contentPrincipal,
+        "persist-data-on-shutdown"
+      ) === Services.perms.ALLOW_ACTION
+    );
+  },
+
+  async toggleSiteData() {
+    const pressed = document.getElementById("identity-popup-allow-sitedata-toggle").toggleAttribute(
+      "pressed"
+    );
+
+    if (pressed) {
+      Services.perms.addFromPrincipal(
+        gBrowser.contentPrincipal,
+        "persist-data-on-shutdown",
+        Services.perms.ALLOW_ACTION,
+        Services.perms.EXPIRE_NEVER
+      );
+    } else {
+      Services.perms.removeFromPrincipal(gBrowser.contentPrincipal, "persist-data-on-shutdown");
+    }
+  },
+
   /**
    * Handler for mouseclicks on the "More Information" button in the
    * "identity-popup" panel.
@@ -1255,6 +1285,8 @@ var gIdentityHandler = {
     this._identityPopupContentOwner.textContent = owner;
     this._identityPopupContentSupp.textContent = supplemental;
     this._identityPopupContentVerif.textContent = verifier;
+
+    this.refreshSiteData();
   },
 
   setURI(uri) {
diff --git a/browser/base/content/browser-trustPanel.js b/browser/base/content/browser-trustPanel.js
index 3b692ef7bf..3673c1ff69 100644
--- a/browser/base/content/browser-trustPanel.js
+++ b/browser/base/content/browser-trustPanel.js
@@ -345,6 +345,9 @@ class TrustPanel {
       document
         .getElementById("trustpanel-toggle")
         .addEventListener("click", () => this.#toggleTrackingProtection());
+      document
+        .getElementById("identity-popup-allow-sitedata-button")
+        .addEventListener("click", () => this.#toggleCookiesAllowed());
       document
         .getElementById("identity-popup-remove-cert-exception")
         .addEventListener("click", () => this.#removeCertException());
@@ -445,6 +448,30 @@ class TrustPanel {
     BrowserCommands.reload();
   }
 
+  #toggleCookiesAllowed() {
+    const pressed = document.getElementById("identity-popup-allow-sitedata-toggle").toggleAttribute(
+      "pressed"
+    );
+
+    if (pressed) {
+      Services.perms.addFromPrincipal(
+        gBrowser.contentPrincipal,
+        "persist-data-on-shutdown",
+        Services.perms.ALLOW_ACTION,
+        Services.perms.EXPIRE_NEVER
+      );
+    } else {
+      Services.perms.removeFromPrincipal(gBrowser.contentPrincipal, "persist-data-on-shutdown");
+    }
+  }
+
+  #areCookiesAllowed() {
+    return Services.perms.testExactPermissionFromPrincipal(
+        gBrowser.contentPrincipal,
+        "persist-data-on-shutdown"
+      ) === Services.perms.ALLOW_ACTION;
+  }
+
   async showPopup(opts = {}) {
     // Avoid flicker between the mouseup and panel shown by manually
     // setting open attribute.
@@ -839,6 +866,9 @@ class TrustPanel {
       { host: this.#displayHost }
     );
 
+    let toggleCookies = document.getElementById("identity-popup-allow-sitedata-toggle");
+    toggleCookies.toggleAttribute("pressed", this.#areCookiesAllowed());
+
     let hostElement = document.getElementById("trustpanel-popup-host");
     hostElement.setAttribute("value", this.#displayHost);
     hostElement.setAttribute("tooltiptext", this.#displayHost);
@@ -878,6 +908,11 @@ class TrustPanel {
       "hidden",
       isPrivate
     );
+    this.#updateAttribute(
+      document.getElementById("trustpanel-allow-cookies-footer"),
+      "hidden",
+      isPrivate
+    );
     if (!isPrivate) {
       try {
         let baseDomain = SiteDataManager.getBaseDomainFromHost(this.#uri.host);
diff --git a/browser/components/controlcenter/content/identityPanel.inc.xhtml b/browser/components/controlcenter/content/identityPanel.inc.xhtml
index 7135a8cb0d..ff6b46d5dd 100644
--- a/browser/components/controlcenter/content/identityPanel.inc.xhtml
+++ b/browser/components/controlcenter/content/identityPanel.inc.xhtml
@@ -98,6 +98,11 @@
         <toolbarbutton id="identity-popup-clear-sitedata-button"
                 data-l10n-id="identity-clear-site-data"
                 class="subviewbutton"/>
+        <toolbarbutton id="identity-popup-allow-sitedata-button"
+                class="subviewbutton">
+          <label data-l10n-id="identity-allow-site-data" flex="1"></label>
+          <html:moz-toggle id="identity-popup-allow-sitedata-toggle" style="pointer-events: none;"></html:moz-toggle>
+        </toolbarbutton>
       </vbox>
       <vbox id="identity-popup-jit-footer"
             class="panel-subview-footer-button">
diff --git a/browser/components/controlcenter/content/trustPanel.inc.xhtml b/browser/components/controlcenter/content/trustPanel.inc.xhtml
index 000e6bd625..4ef1b460af 100644
--- a/browser/components/controlcenter/content/trustPanel.inc.xhtml
+++ b/browser/components/controlcenter/content/trustPanel.inc.xhtml
@@ -75,6 +75,13 @@
           <toolbarbutton data-l10n-id="trustpanel-clear-cookies-button" id="trustpanel-clear-cookies-button" class="subviewbutton subviewbutton-nav"></toolbarbutton>
         </vbox>
 
+        <vbox id="trustpanel-allow-cookies-footer">
+          <toolbarbutton id="identity-popup-allow-sitedata-button" class="subviewbutton">
+            <label data-l10n-id="identity-allow-site-data" flex="1"></label>
+            <html:moz-toggle id="identity-popup-allow-sitedata-toggle" style="pointer-events: none;"></html:moz-toggle>
+          </toolbarbutton>
+        </vbox>
+
         <vbox id="trustpanel-enable-jit-footer">
           <toolbarbutton id="trust-panel-enable-jit-button" class="subviewbutton">
             <label data-l10n-id="panel-enable-jit" flex="1"/>
